python/src/prima/examples/cobyla/cobyla_example.py

```diff
@@ -0,0 +1,322 @@
+
+    constr_loc = np.zeros(m)
+    if f0 is not None and constr0 is not None and all(np.isfinite(x)):
+        f = f0
+        constr_loc = constr0
+    else:
+        # Replace all NaNs with 0s and clip all -inf/inf to -floatmax/floatmax
+        np.nan_to_num(x, copy=False, neginf=-np.finfo(float).max, posinf=np.finfo(float).max)
+        f, constr_loc, cstrv_loc = evaluate(calcfc, x)
+    
+    if rhobeg is not None:
+        rhobeg_loc = rhobeg
+    elif rhoend is not None and np.isfinite(rhoend) and rhoend > 0:
+        rhobeg_loc = max(10*rhoend, RHOBEG_DEFAULT)
+    else:
+        rhobeg_loc = RHOBEG_DEFAULT
+
+    if rhoend is not None:
+        rhoend_loc = rhoend
+    elif rhobeg_loc > 0:
+        rhoend_loc = max(EPS, min(0.1*rhobeg_loc, RHOEND_DEFAULT))
+    else:
+        rhoend_loc = RHOEND_DEFAULT
+
+    maxfun = maxfun if maxfun is not None else MAXFUN_DIM_DEFAULT*n
+
+    if eta1 is not None:
+        eta1_loc = eta1
+    elif eta2 is not None and 0 < eta2 < 1:
+        eta1_loc = max(EPS, eta2/7)
+    else:
+        eta1_loc = 0.1
+
+    if eta2 is not None:
+        eta2_loc = eta2
+    else:
+        if 0 < eta1_loc < 1:
+            eta2_loc = (eta1_loc + 2)/3
+        else:
+            eta2_loc = 0.7
+
+    if maxhist is not None:
+        maxhist_loc = maxhist
+    else:
+        maxhist_loc = max(maxfun, n+2, MAXFUN_DIM_DEFAULT*n)
+
+    # preprocess
+
+    '''
+    ! Further revise MAXHIST_LOC according to MAXHISTMEM, and allocate memory for the history.
+    ! In MATLAB/Python/Julia/R implementation, we should simply set MAXHIST = MAXFUN and initialize
+    ! CHIST = NaN(1, MAXFUN), CONHIST = NaN(M, MAXFUN), FHIST = NaN(1, MAXFUN), XHIST = NaN(N, MAXFUN)
+    ! if they are requested; replace MAXFUN with 0 for the history that is not requested.
+    call prehist(maxhist_loc, n, present(xhist), xhist_loc, present(fhist), fhist_loc, &
+        & present(chist), chist_loc, m, present(conhist), conhist_loc)
+    '''
+
+    # call coblyb, which performs the real calculations
+    x, f, nf, info = cobylb(calcfc, iprint, maxfilt, maxfun, ctol, cweight, eta1_loc, eta2_loc,
+      ftarget, gamma1, gamma2, rhobeg_loc, rhoend_loc, constr_loc, f, x, 
+      cstrv_loc)
+    
+    # Write the outputs.
+    # TODO: Pick up from here, to finish writing the cobyla top level routine, and then we can start testing
+    # We'll be able to test up until the trust region calls, until I hear back from Zaikun about how to handle them
+    return x, f
+
+
+if __name__ == "__main__":
+    # n_chebyquad = 6
+    # x_chebyquad = [i/n_chebyquad for i in range(1, n_chebyquad+1)]
+    m = 0
+    f = 0
+    cstrv = 0
+    # x, f = cobyla(calcfc_chebyquad, m, x_chebyquad, f, cstrv)
+    # print(x, f)
+    x_hexagon = np.zeros(9) + 2
+    m_hexagon = 14
+    f = 0
+    cstrv = 0  # Should this b e a length 14 array? Of inf?
+    x, f = cobyla(calcfc_hexagon, m_hexagon, x_hexagon, f, cstrv)
```

Comment:
in my C-python binding I packed the tuple of results in a Python dataclass
that's convenient because you can add more stuff and keep the same api (I also return the number of function calls)
it's also somewhat compatible with what scipy returns

Reply:
vSorry, this part is incorrect. `cstrv` should be an output. This is specified in the comments, and also indicated by the `intent`: 

- An `intent(in)` argument is an input
- An `intent(out)` argument is an output
- An `intent(inout)` argument is both an input and an output, the value of which will be changed by the procedure / subroutine